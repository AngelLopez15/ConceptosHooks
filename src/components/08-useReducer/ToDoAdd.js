import React from 'react'

import { useForm } from '../../hooks/useForm'

export const ToDoAdd = ({handleAddToDo}) => {
    
    const [{description}, handleInputChange, reset] = useForm({
        description:''
    })
    
    const handleSubmit = (e) =>{
        e.preventDefault()
        if (description.trim().length <= 1) {
            return
        }

        const newToDo = {
            id:new Date().getTime(),
            desc:description,
            done:false
        }

        handleAddToDo(newToDo)
        reset()
    }

    return (
        <>
            <p className="h1">Agregar tarea</p>
            <hr />
            <form onSubmit={handleSubmit}>
                <div className="row mb-4">  
                    <div className="col-12 col-sm-6 mb-2">
                        <input 
                            type="text"
                            name="description"
                            placeholder="comprar ..."
                            autoComplete="off"
                            className="form-control"
                            value={description}
                            onChange={handleInputChange}
                        />
                    </div>
                    <div className="col-12 col-sm-6">
                    <button
                        type="submit"
                        className="btn btn-success btn-block"
                    >Agregar</button>
                    </div>
                </div>
            </form>
        </>
    )
}
