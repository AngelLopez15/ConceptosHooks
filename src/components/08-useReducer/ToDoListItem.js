import React from 'react'

export const ToDoListItem = ({toDo,index,handleDelete,handleToggle}) => {
    return (
        <li
            key={toDo.id}
            className="list-group-item"
        >
        <p className={`mb-0 ${toDo.done && 'complete'}`}
        onClick={()=>handleToggle(toDo.id)}
        >{index+1}. {toDo.desc}</p>
        <button
            className="btn btn-danger btn-block delate"
            onClick={()=>handleDelete(toDo.id)}
        >Borrar</button>
        </li>
    )
}
