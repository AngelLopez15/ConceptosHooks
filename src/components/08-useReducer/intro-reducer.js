// definiendo el estado inicial
// el nombre de la funcion puede tener cualquier nombre

const initialState = [{
    id:1,
    todo:'Comprar el pan',
    done:false,
}]

// creando la funcion reducer
// debemos enviar como argumentos el state y una accion
// las acciones son las que van a modificar el state
const todoReducer = (state=initialState, action)=>{
    // Esta es la funcion que no pude ser asincrona ni usar otras funciones
    // debe ser capaz de hacer todo con su state y su accion
    //  manejando la logica para agregar una nueva tarea
    // Ponemos un ternario para que si la accion tiene type haga el if si no que no haga nada.
    // Aunque lo mas facil es no ejecutar la funcion hasta que no tenga agl 
    if (action?.type === 'agregar') {
        // este es el nuevo state que vamos a retornar
        return [...state, action.payload]
    }

    // siempre debemos regresar un estado aunque no hagamosnada 
    return state
}

let todoList = todoReducer() 

// Para agregar mas tareas

const newTodo= [{
    id:2,
    todo:'Comprar el pan',
    done:false,
}]

// creando la accion
const agregarTodoAction = {
    //es un estandar que la propiedad que le dice al reducer que tipo de accion es
    // se llame type 
    type:'agregar',
    // es un estandar que el nuevo elemento se llame payload
    payload:newTodo
}

todoList = todoReducer(todoList, agregarTodoAction)




console.log(todoList)