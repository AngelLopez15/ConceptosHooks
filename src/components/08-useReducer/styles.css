p{
    cursor: pointer;
    display: inline-block;
    padding-right: 4rem;
    padding-bottom: 1rem;
}


.complete{
    text-decoration: line-through;
}