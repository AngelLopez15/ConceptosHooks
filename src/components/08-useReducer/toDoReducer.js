// creando el reducer en un archivo aparte por que 
// suele suceder que crece mucho esta funcion

export const todoReducer = (state = [], action) => {
    
    switch (action.type) {
        case 'add':
            return [...state, action.payload]
        case 'delete':
            // Buscamos todos los toDos que sean diferentes al id del toDo
            // del payload. Ose el action.payload es igual al toDo.id que no queremos
            return state.filter(toDo=>toDo.id !== action.payload)
        case 'toggle':
            return state.map(toDo=>
                (toDo.id===action.payload)
                    ?{...toDo, done: !toDo.done}
                    :toDo
                )
        case 'toggle-optional':
            return state.map(toDo=>{
                if (toDo.id===action.payload) {
                    return{
                        ...toDo,
                        done:!toDo.done
                    }
                } else {
                    return toDo
                }
            })
        default:
            return state
    }
}