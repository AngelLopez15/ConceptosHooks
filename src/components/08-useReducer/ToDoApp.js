import React, { useReducer, useEffect } from 'react'
import { todoReducer } from './toDoReducer'
import { useForm } from '../../hooks/useForm'

import './styles.css'

// creando el estado inicial
// const initialState = [{
//     id:new Date().getTime(),
//     desc:'Aprender React',
//     done:false
// }]

const init = () =>{
    
    return JSON.parse(localStorage.getItem('toDos')) || []

    // return [{
    //     id:new Date().getTime(),
    //     desc:'Aprender React',
    //     done:false
    // }]
}

export const ToDoApp = () => {

    // declarando el reducer
    // reducer es la funcion reducer
    // initialstate es el estado inicial
    // init es una funcion para inicializar el state en caso de que el state
    // en caso de que el state sea procesado o haga varias acciones. Ya que 
    // cada vez que el componenete se vuelva a ejecutar al ser una funcion externa
    // lo hace mas ligero y rapido. Incluso puede memorizar el resultado para no
    // volverlo a ejecutar en caso de ser necesario.

    // dispatch sirve para disparar las acciones para el reducer.

    // const [state, dispatch] = useReducer(reducer, initialState, init)

    const [toDos, dispatch] = useReducer(todoReducer, [], init)
    
    const [{description}, handleInputChange, reset] = useForm({
        description:''
    })

    // Para guardar los datos en localStorage
    useEffect(() => {
        localStorage.setItem('toDos', JSON.stringify(toDos))
    }, [toDos])

    // para borrar una tarea
    const handleDelete=(toDoId)=>{

        const action = {
            type:'delete',
            payload:toDoId
        }

        dispatch(action)
    }

    const handleToogle=(toDoId)=>{
        dispatch({
            type:'toggle',
            payload:toDoId
        })
    }

    // aÃ±ade una nueva tarea
    const handleSubmit = (e) =>{
        e.preventDefault()
        if (description.trim().length <= 1) {
            return
        }

        const newToDo = {
            id:new Date().getTime(),
            desc:description,
            done:false
        }

        const action = {
            type:'add',
            payload:newToDo
        }

        dispatch(action)
        reset()
    }
    

    return (
        <div className="container mt-4">
            <h1>ToDoList App</h1>
            <hr />
            <p className="h1">Agregar tarea</p>
            <hr />
            <form onSubmit={handleSubmit}>
                <div className="row mb-4">  
                    <div className="col-12 col-sm-6 mb-2">
                        <input 
                            type="text"
                            name="description"
                            placeholder="comprar ..."
                            autoComplete="off"
                            className="form-control"
                            value={description}
                            onChange={handleInputChange}
                        />
                    </div>
                    <div className="col-12 col-sm-6">
                    <button
                        type="submit"
                        className="btn btn-success btn-block"
                    >Agregar</button>
                    </div>
                </div>
            </form>
            <div className="row mb-2 justify-content-center">
                <div className="col-12">
                    <ul className="list-group list-group-flush">
                        {
                            toDos.map((toDo, i) => (
                                <li
                                    key={toDo.id}
                                    className="list-group-item"
                                >
                                <p className={`mb-0 ${toDo.done && 'complete'}`}
                                onClick={()=>handleToogle(toDo.id)}
                                >{i+1}. {toDo.desc}</p>
                                <button
                                    className="btn btn-danger btn-block delate"
                                    onClick={()=>handleDelete(toDo.id)}
                                >Borrar</button>
                                </li>
                            ))
                        }
                    </ul>
                </div>
            </div>
        </div>
    )
}
