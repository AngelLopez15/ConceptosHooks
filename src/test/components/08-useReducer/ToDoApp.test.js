import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { ToDoApp } from '../../../components/08-useReducer/ToDoApp';
import { act } from '@testing-library/react';
import { demoTodos } from '../../fixtures/demosToDos';


describe('Pruebas en <ToDoApp />', () => {

  configure({adapter: new Adapter()});

  const wrapper = shallow(<ToDoApp />)

  Storage.prototype.setItem=jest.fn(()=>{})

  test('Debe de mostarse correctamente', () => {
    
    expect(toJson(wrapper)).toMatchSnapshot();

  })

  test('Debe de agregar un ToDo', () => {
    
    const wrapper = mount(<ToDoApp />)

    act(()=>{
      wrapper.find('ToDoAdd').prop('handleAddToDo')(demoTodos[0])
      wrapper.find('ToDoAdd').prop('handleAddToDo')(demoTodos[1])
    })

    expect(wrapper.find('h1').text().trim()).toBe('ToDoList App')

    expect(localStorage.setItem).toHaveBeenCalledTimes(2)
    
  });

  test('Debe de eliminar un ToDo', () => {
    
    wrapper.find('ToDoAdd').prop('handleAddToDo')(demoTodos[0])
    wrapper.find('ToDoList').prop('handleDelete')(demoTodos[0].id)


  });

});