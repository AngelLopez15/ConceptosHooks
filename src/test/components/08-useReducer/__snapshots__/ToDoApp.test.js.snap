// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <ToDoApp /> Debe de mostarse correctamente 1`] = `
<div
  className="container mt-4"
>
  <h1>
    ToDoList App
  </h1>
  <hr />
  <ToDoAdd
    handleAddToDo={[Function]}
  />
  <div
    className="row mb-2 justify-content-center"
  >
    <div
      className="col-12"
    >
      <ToDoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        toDos={Array []}
      />
    </div>
  </div>
</div>
`;
